File: StriverA2Z/C++/3_ArraysQuestion/3.2_Medium/leaders_in_array.cpp
#include<bits/stdc++.h>

using namespace std;

void leader(int arr[], int n) {
    vector<int> ans;
    int maximum = INT_MIN;
    for (int i = n - 1; i >= 0; i--) {
        if (arr[i] > maximum) ans.push_back(arr[i]);
        maximum = max(maximum, arr[i]);
    }

    // Printing
    for (int i: ans) {
        cout << i << " ";
    }
}

int main() {
    int arr[] = {10, 22, 12, 3, 0, 6};
    leader(arr, sizeof(arr) / sizeof(arr[0]));
    return 0;
}

File: StriverA2Z/Java/Arrays/Medium/LeadersInArray.java
import java.util.ArrayList;

public class LeadersInArray {
    public static void main(String[] args) {
        int[] arr = {7, 10, 4, 10, 6, 5, 2};
        leader(arr);
    }

    static void leader(int[] arr) {
        ArrayList<Integer> ans = new ArrayList<>();
        int max = Integer.MIN_VALUE;
        for (int i = arr.length - 1; i >= 0; i--) {
            if (arr[i] > max) ans.add(arr[i]);
            max = Math.max(max, arr[i]);
        }

        // printing
        for (Integer i : ans) {
            System.out.print(i + " ");
        }
    }
}